src/feature-reef/enter_exit.py

```
@@ -0,0 +1,43 @@
+# Dunders __enter__ and __exit__
+
+from __future__ import annotations
+
+
+def handle_error(exception):
+    ...
+
+
+class Context:
+    def __enter__(self) -> None:
+        print("entering with block")
+
+    def __exit__(
+            self, exception_type, exception_value,
+            exception_traceback) -> None:
+        print("exiting with block")
+
+        if (exception_value is not None):
+            print(f"exception '{exception_type}' "
+                  f"occurred at line {exception_traceback.tb_lineno}: "
+                  f"{exception_value}")
+
+            print("handling error")
+            handle_error(exception_value)
+
+
+def main() -> None:
+    doc = Context()
+
+    with doc:
+        print("\nExecuting fancy function\n")
+
+    try:
+        with doc:
+            raise Exception("error")
+
+    except Exception:
+        pass
```

not entirely sure what this is for

---

It throws an exception which is then handled by the `__exit__` function

---

yes but why the try/except if `__exit__` handles the exception

---

The exception still stops the program

---

> The exception still stops the program even though it is passed into the arguments of the `__exit__`

Oh yeah, that's because `__exit__` needs to return `True` to suppress the error.

---

In which case I think the easiest way to handle it would be to do something like
```diff
-print("handling error")
-handle_error(exception_value)
+print("ignoring error")
+return True
```
and
```diff
-    try:
-        with Context() as ctx:
-            raise Exception("error")
-
-    except Exception:
-        pass
+    with Context() as ctx:
+        raise Exception("error")
```